python/shotgun_model/data_handler.py

```diff
@@ -0,0 +1,473 @@
+
+    Each QT model typically sublcasses this in order to create
+    a specific data handler. These methods need to be implemented
+    by all deriving classes:
+
+    - generate_data_request - called by the model when it needs
+      additional data to be loaded from shotgun. The data handler
+      formulates the exact request to be sent out to the server.
+
+    - update_data - the counterpart of generate_data_request: this
+      is called when the requested shotgun data is returned and
+      needs to be inserted into the data structure.
+
+    Data returned back from this class to the Model layer
+    is always sent as ShotgunDataItem object to provide a full
+    encapsulation around the internals of this class.
+    """
+    # version of binary format - increment this whenever changes
+    # are made which renders the cache files non-backwards compatible.
+    FORMAT_VERSION = 27
+
+    # constants for updates
+    (UPDATED, ADDED, DELETED) = range(3)
+
+    # internal constants for serialization performance
+    (CACHE_BY_UID, CACHE_CHILDREN, IS_LEAF, UID, PARENT, FIELD, SG_DATA) = range(7)
+
+    def __init__(self, cache_path, parent):
+        """
+        :param cache_path: Path to cache file location
+        :param parent: Parent QT object
+        """
+        super(ShotgunDataHandler, self).__init__(parent)
+        # keep a handle to the current app/engine/fw bundle for convenience
+        self._bundle = sgtk.platform.current_bundle()
+        # the path to the cache file
+        self._cache_path = cache_path
+        # data in cache
+        self._cache = None
+
+    def __repr__(self):
+        """
+        Create a string representation of this instance
+        :returns: A string representation of this instance
+        """
+        return "<%s@%s (%d items)>" % (
+            self.__class__.__name__,
+            self._cache_path,
+            len(self._cache[self.CACHE_BY_UID])
+        )
+
+    def _create_clear_cache(self):
+        """
+        Helper method - initializes a clear cache.
+
+        :returns: new cache dictionary
+        """
+        return {
+            self.CACHE_CHILDREN: {},  # hierarchy
+            self.CACHE_BY_UID: {},    # uid-based lookup
+            self.UID: None            # the uid of the root is None
+        }
+
+    def _clear_cache(self):
+        """
+        Sets up an empty cache in memory
+        """
+        self._cache = self._create_clear_cache()
+
+    def is_cache_available(self):
+        """
+        Returns true if the cache exists on disk, false if not.
+
+        :returns: boolean to indicate if cache exists on disk
+        """
+        return os.path.exists(self._cache_path)
+
+    def is_cache_loaded(self):
+        """
+        Returns true if the cache has been loaded into memory, false if not.
+
+        :returns: boolean to indicate if cache is loaded
+        """
+        return self._cache is not None
+
+    @log_timing
+    def remove_cache(self):
```

Comment:
I agree with Josh. I find the state transitions confusing. The documentation is also confusing. By looking at the method's name and then the code I would have expected `is_cached_loaded` to return False after the call to `remove_cache`. However, this is not true, as `load_cache` is called right at the end.